src/context/cdf_context.rs

```diff
@@ -601,6 +601,10 @@ trait CDFContextLogOps: CDFContextLogSize {
       let capacity = log.data.capacity();
       debug_assert!(new_len <= capacity);
       let dst = log.data.as_mut_ptr().add(len);
+      // This reads beyond the end of `cdf`.
+      // Since it's part of a larger data structure, the out-of-bounds data is valid.
```

Comment:
The larger data structure is `struct CDFContext` and we should have enough language features available now to rework this module.

Reply:
I have completed the rework in #3169 and #3172, please take a look and rebase.